File: sdk/swift/Sources/Litebase/TableQuery.swift
import Foundation

/// Fluent table accessor: CRUD + realtime for one table name.
public struct TableQuery: Sendable {
    private let client: LitebaseClient
    private let table: String

    init(client: LitebaseClient, table: String) {
        self.client = client
        self.table = table
    }

    public func select(
        where whereClause: [String: WhereValue]? = nil,
        sort: String? = nil,
        limit: Int? = nil,
        offset: Int? = nil
    ) async throws -> ListResult {
        var items: [URLQueryItem] = []
        if let limit { items.append(URLQueryItem(name: "limit", value: String(limit))) }
        if let offset { items.append(URLQueryItem(name: "offset", value: String(offset))) }
        if let sort { items.append(URLQueryItem(name: "sort", value: sort)) }
        if let whereClause {
            for (k, v) in whereClause {
                items.append(URLQueryItem(name: "where[\(k)]", value: v.queryString))
            }
        }

        var path = "/api/\(encodeURIComponent(table))"
        if !items.isEmpty {
            var components = URLComponents()
            components.queryItems = items
            if let q = components.percentEncodedQuery {
                path += "?\(q)"
            }
        }

        let env: ListEnvelope = try await client.requestJSON(method: "GET", path: path)
        return ListResult(data: env.data, meta: env.meta)
    }

    public func get(_ id: String) async throws -> [String: JSONValue] {
        let path = "/api/\(encodeURIComponent(table))/\(encodeURIComponent(id))"
        let env: DataEnvelope<[String: JSONValue]> = try await client.requestJSON(
            method: "GET",
            path: path
        )
        return env.data
    }

    public func get(_ id: Int) async throws -> [String: JSONValue] {
        try await get(String(id))
    }

    public func insert(_ row: [String: JSONValue]) async throws -> [String: JSONValue] {
        let body = try JSONEncoder().encode(row)
        let path = "/api/\(encodeURIComponent(table))"
        let env: DataEnvelope<[String: JSONValue]> = try await client.requestJSON(
            method: "POST",
            path: path,
            body: body
        )
        return env.data
    }

    public func update(_ id: String, patch: [String: JSONValue]) async throws -> [String: JSONValue] {
        let body = try JSONEncoder().encode(patch)
        let path = "/api/\(encodeURIComponent(table))/\(encodeURIComponent(id))"
        let env: DataEnvelope<[String: JSONValue]> = try await client.requestJSON(
            method: "PATCH",
            path: path,
            body: body
        )
        return env.data
    }

    public func update(_ id: Int, patch: [String: JSONValue]) async throws -> [String: JSONValue] {
        try await update(String(id), patch: patch)
    }

    public func delete(_ id: String) async throws -> [String: JSONValue] {
        let path = "/api/\(encodeURIComponent(table))/\(encodeURIComponent(id))"
        let env: DataEnvelope<[String: JSONValue]> = try await client.requestJSON(
            method: "DELETE",
            path: path
        )
        return env.data
    }

    public func delete(_ id: Int) async throws -> [String: JSONValue] {
        try await delete(String(id))
    }

    @discardableResult
    public func subscribe(
        rowId: String? = nil,
        handler: @escaping SubscribeHandler
    ) -> Unsubscribe {
        client.subscribeTable(table: table, rowId: rowId, handler: handler)
    }

    /// Awaitable subscribe — resolves when the server acknowledges (`subscribed` frame).
    public func subscribeReady(
        rowId: String? = nil,
        timeoutMs: Int = 5000,
        handler: @escaping SubscribeHandler
    ) async throws -> Unsubscribe {
        try await client.subscribeTableReady(
            table: table,
            rowId: rowId,
            timeoutMs: timeoutMs,
            handler: handler
        )
    }
}

File: sdk/flutter/lib/src/table_query.dart
import 'dart:async';
import 'dart:convert';

import 'client.dart';
import 'models.dart';
import 'realtime_url.dart';

/// Fluent table accessor: CRUD + realtime for one table name.
class TableQuery {
  final LoomupClient _client;
  final String table;

  TableQuery(this._client, this.table);

  /// List rows with optional filters.
  ///
  /// Boolean `where` values are encoded as SQLite `0`/`1`.
  Future<ListResult> select({
    Map<String, Object?>? where,
    Map<String, Map<String, Object?>>? filter,
    List<String>? select,
    String? sort,
    int? limit,
    int? offset,
    String? cursor,
  }) async {
    final params = <String, String>{};
    if (cursor != null) params['cursor'] = cursor;
    if (limit != null) params['limit'] = limit.toString();
    if (offset != null) params['offset'] = offset.toString();
    if (sort != null) params['sort'] = sort;
    if (select != null && select.isNotEmpty) {
      params['select'] = select.join(',');
    }
    if (where != null) {
      for (final entry in where.entries) {
        final v = entry.value;
        if (v is bool) {
          params['where[${entry.key}]'] = v ? '1' : '0';
        } else if (v != null) {
          params['where[${entry.key}]'] = v.toString();
        }
      }
    }
    if (filter != null) {
      for (final field in filter.entries) {
        for (final operation in field.value.entries) {
          final wireOperation = switch (operation.key) {
            'isNull' => 'is_null',
            'startsWith' => 'starts_with',
            _ => operation.key,
          };
          final value = operation.value;
          if (value == null) continue;
          params['filter[${field.key}][$wireOperation]'] = value is Iterable
              ? value.join(',')
              : value.toString();
        }
      }
    }

    var path = '/api/${encodeUriComponent(table)}';
    if (params.isNotEmpty) {
      final q = params.entries
          .map((e) =>
              '${Uri.encodeQueryComponent(e.key)}=${Uri.encodeQueryComponent(e.value)}')
          .join('&');
      path = '$path?$q';
    }

    final json = await _client.requestJson('GET', path);
    final dataRaw = json['data'];
    final data = dataRaw is List
        ? dataRaw
            .whereType<Map>()
            .map((e) => Map<String, dynamic>.from(e))
            .toList()
        : <Map<String, dynamic>>[];
    final metaRaw = json['meta'];
    final meta = metaRaw is Map<String, dynamic>
        ? ListMeta.fromJson(metaRaw)
        : const ListMeta(limit: 0, offset: 0, total: 0);
    return ListResult(data: data, meta: meta);
  }

  Future<Map<String, dynamic>> get(Object id) async {
    final path =
        '/api/${encodeUriComponent(table)}/${encodeUriComponent(id.toString())}';
    final json = await _client.requestJson('GET', path);
    return Map<String, dynamic>.from(json['data'] as Map);
  }

  Future<Map<String, dynamic>> insert(Map<String, dynamic> row) async {
    final path = '/api/${encodeUriComponent(table)}';
    final json = await _client.requestJson(
      'POST',
      path,
      body: row,
    );
    return Map<String, dynamic>.from(json['data'] as Map);
  }

  Future<Map<String, dynamic>> update(
    Object id,
    Map<String, dynamic> patch,
  ) async {
    final path =
        '/api/${encodeUriComponent(table)}/${encodeUriComponent(id.toString())}';
    final json = await _client.requestJson(
      'PATCH',
      path,
      body: patch,
    );
    return Map<String, dynamic>.from(json['data'] as Map);
  }

  Future<Map<String, dynamic>> delete(Object id) async {
    final path =
        '/api/${encodeUriComponent(table)}/${encodeUriComponent(id.toString())}';
    final json = await _client.requestJson('DELETE', path);
    return Map<String, dynamic>.from(json['data'] as Map);
  }

  /// Subscribe to change events. Returns an unsubscribe function.
  Unsubscribe subscribe(SubscribeHandler handler, {String? rowId}) {
    return _client.subscribeTable(table, handler, rowId: rowId);
  }

  /// Subscribe and wait until the server acknowledges (`subscribed` frame).
  /// Prefer this when the next statement mutates data.
  Future<Unsubscribe> subscribeReady(
    SubscribeHandler handler, {
    String? rowId,
    int timeoutMs = 5000,
  }) {
    return _client.subscribeTableReady(
      table,
      handler,
      rowId: rowId,
      timeoutMs: timeoutMs,
    );
  }

  /// Dart-idiomatic stream of change events. Cancelling the subscription
  /// unsubscribes from the table/row channel.
  Stream<ChangeEvent> changes({String? rowId}) {
    late StreamController<ChangeEvent> controller;
    Unsubscribe? unsub;
    controller = StreamController<ChangeEvent>(
      onListen: () {
        unsub = subscribe((ev) {
          if (!controller.isClosed) controller.add(ev);
        }, rowId: rowId);
      },
      onCancel: () {
        unsub?.call();
        unsub = null;
      },
    );
    return controller.stream;
  }
}

/// Encode body map as UTF-8 JSON bytes (used by client tests / helpers).
List<int> encodeJsonBody(Object? body) => utf8.encode(jsonEncode(body));
